apps/devhub/tests/test_utils.py

```diff
@@ -0,0 +1,373 @@
+                          'context': ('x', 'y', 'z'),
+                          'file': 'foo.js'},
+                         {},
+                         {'ignored': True,
+                          'matched': True})
+
+            self.compare({'id': ('a', 'b', 'c'),
+                          'signing_severity': severity,
+                          'ignore_duplicates': False,
+                          'context': ('x', 'y', 'z'),
+                          'file': 'foo.js'},
+                         {},
+                         {'ignored': False,
+                          'matched': True})
+
+        # Other severities are ignored only when flagged as ignorable.
+        for severity in 'medium', 'high':
+            self.compare({'id': ('a', 'b', 'c'),
+                          'signing_severity': severity,
+                          'context': ('x', 'y', 'z'),
+                          'file': 'foo.js'},
+                         {},
+                         {'ignored': False,
+                          'matched': True})
+
+            self.compare({'id': ('a', 'b', 'c'),
+                          'signing_severity': severity,
+                          'ignore_duplicates': True,
+                          'context': ('x', 'y', 'z'),
+                          'file': 'foo.js'},
+                         {},
+                         {'ignored': True,
+                          'matched': True})
+
+        # Check that unchecked properties don't matter.
+        self.compare({'id': ('a', 'b', 'c'),
+                      'signing_severity': 'low',
+                      'context': ('x', 'y', 'z'),
+                      'file': 'foo.js'},
+                     {'thing': 'stuff'},
+                     {'ignored': True,
+                      'matched': True})
+
+        # Non-signing messages should be matched, but not annotated for
+        # ignorability.
+        self.compare({'id': ('a', 'b', 'c'),
+                      'context': ('x', 'y', 'z'),
+                      'file': 'foo.js'},
+                     {},
+                     {'matched': True})
+
+    def test_non_matching_messages(self):
+        """Test that messages which should match, don't."""
+
+        message = {'id': ('a', 'b', 'c'),
+                   'signing_severity': 'low',
+                   'context': ('x', 'y', 'z'),
+                   'context_data': {'function': 'foo_bar'},
+                   'file': 'foo.js'}
+
+        self.compare(message, {'id': ('d', 'e', 'f')}, {})
+        self.compare(message, {'signing_severity': 'high'}, {})
+        self.compare(message, {'context': ('a', 'b', 'c')}, {})
+        self.compare(message, {'context_data': {}}, {})
+        self.compare(message, {'file': 'these-are-not-the-droids.js'}, {})
+
+        # Token non-signing message.
+        self.compare({'id': ('a', 'b', 'c'),
+                      'context': ('x', 'y', 'z'),
+                      'file': 'foo.js'},
+                     {'id': ()}, {})
```

Comment:
I'm not that happy about this TestCase, it looks well over-engineered, and too clever for a test. If possible, I'd rather you simplify this, even if it means more code at the end. Tests should be as simple as possible for two reasons:
1/ maintainability: it's important that they're as simple as possible to be as understandable as possible, to make sure we understand what is tested, what is expected...
2/ more code means more possible bugs, and we don't want bugs in our test code. Or we need to test this test code.